src/connector/src/sink/pulsar.rs

```diff
@@ -0,0 +1,321 @@
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct PulsarSink {
+    pub config: PulsarConfig,
+    schema: Schema,
+    downstream_pk: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl PulsarSink {
+    pub fn new(config: PulsarConfig, param: SinkParam) -> Self {
+        Self {
+            config,
+            schema: param.schema(),
+            downstream_pk: param.downstream_pk,
+            is_append_only: param.sink_type.is_append_only(),
+        }
+    }
+}
+
+#[async_trait]
+impl Sink for PulsarSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type Writer = PulsarSinkWriter;
+    async fn new_writer(&self, _writer_param: SinkWriterParam) -> Result<Self::Writer> {
+        PulsarSinkWriter::new(
+            self.config.clone(), 
+            self.schema.clone(), 
+            self.downstream_pk.clone(),
+            self.is_append_only,
+        ).await
+    }
+
+    async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
+        // For upsert Pulsar sink, the primary key must be defined.
+        if !self.is_append_only && self.downstream_pk.is_empty() {
+            return Err(SinkError::Config(anyhow!(
+                "primary key not defined for {} pulsar sink (please define in `primary_key` field)",
+                self.config.r#type
+            )));
+        }
+
+        // TODO: validate pulsar connection
+
+        Ok(())
+    }
+}
+
+pub struct PulsarSinkWriter {
+    pulsar: Pulsar<TokioExecutor>,
+    producer: Producer<TokioExecutor>,
+    config: PulsarConfig,
+    schema: Schema,
+    downstream_pk: Vec<usize>,
+    is_append_only: bool,
+    send_future_buffer: VecDeque<SendFuture>,
+}
+
+impl PulsarSinkWriter {
+    pub async fn new(
+        config: PulsarConfig, 
+        schema: Schema, 
+        downstream_pk: Vec<usize>, 
+        is_append_only: bool
+    ) -> Result<Self> {
+        let pulsar = config.common.build_client().await?;
+        let producer = pulsar
+            .producer()
+            .with_topic(&config.common.topic)
+            .build()
+            .map_err(|e| SinkError::Pulsar(anyhow!("Pulsar sink error: {}", e)))
+            .await?;
+        Ok(Self {
+            pulsar,
+            producer,
+            config,
+            schema,
+            downstream_pk,
+            is_append_only,
+            send_future_buffer: VecDeque::new(),
+        })
+    }
+
+    async fn send_message(&mut self, message: Message) -> Result<()> {
+        let mut success_flag = false;
+        let mut connection_err = None;
+
+        for _ in 0..self.config.max_retry_num {
+            match self.producer.send(message.clone()).await {
+                // If the message is sent successfully,
+                // a SendFuture holding the message receipt 
+                // or error after sending is returned
+                Ok(send_future) => {
+                    // Check if send_future_buffer is greater than the preset limit
```

Comment:
batch_size and batch_byte_size can be specified upon building the producer.
See [doc](https://docs.rs/pulsar/latest/pulsar/producer/struct.ProducerOptions.html#structfield.batch_size) and [example](https://github.com/streamnative/pulsar-rs/blob/master/examples/batching.rs#L46C9-L53C11).

Reply:
I suggest to create batch.num and batch.size in the config fields, with the same default value (10k and 1MB) as rdkafka. linger can not be specified in pulsar-rs.

Reply:
One question: since maximum number and size of messages in a batch is handled by the kafka/pulsar client, why should we check the 65536 threshold of buffered DeliveryFuture/SendFuture?

Reply:
> One question: since maximum number and size of messages in a batch is handled by the kafka/pulsar client, why should we check the 65536 threshold of buffered DeliveryFuture/SendFuture?

The batched messages in `rdkafka` or `pulsar` are not essentially the same as the buffered futures in RW's connector.
In general, after a message is successfully sent to the client, we'll have a corresponding future buffered. The reason for this is to ensure the message has eventually been sent to the downstream (i.e., The Kafka Broker) by the `rdkafka` or `pulsar`. Otherwise the status of the messages can not be known without this.
If the messages are eventually delivered, there's nothing to do. If not, we will directly rollback to the latest checkpoint when any error is returned through the process.
Also, when the barrier is triggered by the sink coordinator, we'll group commit all the current buffered future to ensure every message is delivered. If not, also rollback to the latest checkpoint.
This is why we need an extra threshold for the buffered futures, in which case will ensure a efficient interval to check the status of all the sent but not yet delivered messages.